rust/gitxetcore/src/data/remote_shard_interface.rs

```diff
@@ -110,41 +165,84 @@ impl RemoteShardInterface {
             shard_manager
         };
 
-        Ok(Self {
+        Ok(Arc::new(Self {
             config: config.clone(),
+            repo_salt,
             smudge_query_policy: config.smudge_query_policy,
             shard_manager,
             shard_client,
             reconstruction_cache: Mutex::new(LruCache::new(FILE_RECONSTRUCTION_CACHE_SIZE)),
             cas,
             shard_downloads: Arc::new(singleflight::Group::new()),
-        })
+        }))
     }
 
-    pub async fn new_local(shard_manager: Arc<ShardFileManager>) -> Result<Self> {
-        Ok(Self {
-            config: XetConfig::empty(),
-            smudge_query_policy: SmudgeQueryPolicy::LocalOnly,
-            shard_manager: Some(shard_manager),
-            shard_client: None,
-            cas: None,
-            reconstruction_cache: Mutex::new(LruCache::new(FILE_RECONSTRUCTION_CACHE_SIZE)),
-            shard_downloads: Arc::new(singleflight::Group::new()),
-        })
+    pub fn cas(&self) -> Result<Arc<dyn Staging + Send + Sync>> {
+        let Some(cas) = self.cas.clone() else {
+            // Trigger error and backtrace
+            Err(anyhow!("cas requested but has not been configured."))?;
+            unreachable!();
+        };
+
+        Ok(cas)
+    }
+
+    pub fn shard_client(&self) -> Result<Arc<dyn ShardClientInterface>> {
+        let Some(shard_client) = self.shard_client.clone() else {
+            // Trigger error and backtrace
+            Err(anyhow!(
+                "shard_client requested but has not been configured."
+            ))?;
+            unreachable!();
+        };
+
+        Ok(shard_client)
+    }
+
+    pub fn shard_manager(&self) -> Result<Arc<ShardFileManager>> {
+        let Some(shard_manager) = self.shard_manager.clone() else {
+            // Trigger error and backtrace
+            Err(anyhow!(
+                "shard_manager requested but has not been configured."
+            ))?;
+            unreachable!();
+        };
+
+        Ok(shard_manager)
+    }
+
+    pub fn shard_prefix(&self) -> String {
+        self.config.cas.shard_prefix()
+    }
+
+    pub fn shard_cache_dir(&self) -> &Path {
+        &self.config.merkledb_v2_cache
+    }
+
+    pub fn repo_salt(&self) -> Result<RepoSalt> {
+        let Some(salt) = self.repo_salt else {
+            // Trigger error and backtrace
+            Err(anyhow!(
+                "ERROR: Repo salt requested but has not been configured."
+            ))?;
+            unreachable!();
+        };
+        Ok(salt)
     }
 
     pub async fn query_server(
```

Comment:
This struct now supports both smudge query and dedup query so this function name becomes ambiguous. Maybe make this private and rename it?

Reply:
Good catch!

Reply:
Good catch.